src/solve.rs

```diff
@@ -1374,4 +1377,73 @@ mod tests {
 
         println!("{:#?}", solution);
     }
+
+    #[test]
+    fn combines_subtrades_joint_token_pairs() {
+        let usdc = H160(hex!("a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"));
+        let weth = H160(hex!("c02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"));
+
+        // Depending on the order that the subtrades are considered (which is
+        // random because of `HashMap`), it can either sell excess WETH or USDC.
+        // Deal with this by trying until we hit both cases.
+        let mut sold_usdc = false;
+        let mut sold_weth = false;
+
+        while !sold_usdc || !sold_weth {
```

Comment:
Not a fan of this. Hashmap makes no guarantees about order which includes not guaranteeing true randomness which means we could loop forever. I'm wondering if you could refactor the input to that function in a way that makes it deterministic.
For example instead of `&HashMap<(H160, H160), (U256, U256)>` there could be

```rust
struct SingleInput {
    token0: H160,
    token1: H160,
    sell0Buy1Amounts: Option<(U256, U256)>,
    sell1Buy0Amounts: Option<(U256, U256)>,
}
```
(can probably make it nicer)

and the function takes an iterator of these. This should work because the function only ever looks at a pair and the reverse together.

Reply:
> not guaranteeing true randomness which means we could loop forever.

True. I explained this away to myself since this is test code.

> I'm wondering if you could refactor the input to that function in a way that makes it deterministic.

Unfortunately, it uses hash-map look up for this. I will try and see if I can work something out.

Reply:
But aren't the only look ups it does always for token pair and the reverse? So it doesn't need a complete hashmap, just for every token pair both directions. Or maybe you're talking about code that wasn't part of this PR that I didn't see.

Reply:
Yeah, we could refactor it. Or use: https://docs.rs/linked-hash-map/latest/linked_hash_map/

But given that we don't really want to maintain it, I am also fine with the current status quo.

Reply:
> But aren't the only look ups it does always for token pair and the reverse?

Yes. So, in general I think your idea is the right approach. I was just lazy to make the bigger change because it involves adjusting tests.

> Or maybe you're talking about code that wasn't part of this PR that I didn't see.

Yes, later when constructing the clearing prices, that logic also looks at the `split_token_amounts` but ALSO queries both the token pair and its reverse.

In general, I think your idea is the correct one, and the only reason to not implement it is... well laziness 😛.